config.go

```diff
@@ -1362,13 +1363,65 @@ func extractBitcoindRPCParams(bitcoindConfigPath string) (string, string, string
 }
 
 // checkZMQOptions ensures that the provided addresses to use as the hosts for
-// ZMQ rawblock and rawtx notifications are different.
+// ZMQ rawblock and rawtx notifications are valid and different.
 func checkZMQOptions(zmqBlockHost, zmqTxHost string) error {
-	if zmqBlockHost == zmqTxHost {
-		return errors.New("zmqpubrawblock and zmqpubrawtx must be set" +
+	// ZMQ can be used with diferent transports
+	// (http://api.zeromq.org/4-0:zmq-bind). At first, parse the socket URL
+	// and perform generalized comparison.
+	zmqBlockURL, err := url.Parse(zmqBlockHost)
+	if err != nil {
+		return err
+	}
+
+	zmqTxURL, err := url.Parse(zmqTxHost)
+	if err != nil {
+		return err
+	}
+
+	if zmqBlockURL == zmqTxURL {
+		return errors.New("zmqpubrawblock and zmqpubrawtx must be set " +
 			"to different addresses")
 	}
 
+	// TCP transport specific checks.
+	if zmqBlockURL.Scheme == "tcp" && zmqTxURL.Scheme == "tcp" {
+		zmqBlockResolved, err := net.ResolveTCPAddr("tcp", zmqBlockURL.Host)
+		if err != nil {
+			return err
+		}
+		zmqBlockHost, zmqBlockPort, err :=
+			net.SplitHostPort(zmqBlockResolved.String())
+		if err != nil {
+			return err
+		}
+		zmqBlockHostIP := net.ParseIP(zmqBlockHost)
+
+		zmqTxResolved, err := net.ResolveTCPAddr("tcp", zmqTxURL.Host)
+		if err != nil {
+			return err
+		}
+		zmqTxHost, zmqTxPort, err := net.SplitHostPort(zmqTxResolved.String())
+		if err != nil {
+			return err
+		}
+		zmqTxHostIP := net.ParseIP(zmqTxHost)
+
+		if zmqBlockPort == zmqTxPort &&
+			// Protects against one is 127.0.0.1 and the other is ::1
+			((zmqBlockHostIP.IsLoopback() && zmqTxHostIP.IsLoopback()) ||
```

Comment:
This is a bit hard to read. Can you split the cases into two, and ideally return a slightly different error to differentiate? :)

Reply:
Uh, you really want different errors here? The `IsLoopback` is used here just because `Equal` returns `false` in case of `127.0.0.1` and `::1` (basically, because `127.0.0.1` maps to `0:0:0:0:0:ffff:7f00:1`). So it's kinda workaround here.

Reply:
you could do 
```
switch {
// Comment 1
case zmqBlockHostIP.IsLoopback() && zmqTxHostIP.IsLoopback:
     fallthrough

// Comment 2..
case zmqBlockHostIP.Equal(zmqTxHostIP):
     return errors.New("....")
}
```

Reply:
Ah, ok. Done!